blog/_posts/2018-06-04-scalac-profiling.md

```diff
@@ -0,0 +1,1327 @@
+---
+layout: blog-detail
+post-type: blog
+by: Jorge Vicente Cantero
+title: "Profiling and reducing compile times of typeclass derivation"
+---
+
+Today I write about how I've reduced compilation times of a project that uses
+typeclass derivation to create parsers for CLIs (command-line interfaces)
+with Shapeless.
+
+Typeclass derivation makes a heavy use of macros and implicit search to reduce
+boilerplate and reduce code maintenance, but has a high cost on compile
+times. In most of the cases, slow compilations of this common use case in the
+Scala community originate from either an unintentional misuse of a
+macro-based library, or inefficient macro implementations.
+
+Until now, Scala developers could not reduce the cost of the slow compilation
+times because tooling to profile compile times was lacking and knowledge of these issues limited.
+
+My goal in this blog post is to give you a walkthrough on how to find and
+reduce these compile times with `scalac-profiling`. `scalac-profiling` is a
+new Scala Center compiler plugin to complement my recent work on the compiler
+statistics infrastructure merged in 2.12.5. In this guide, I use the plugin
+to speed up a module of [Bloop](https://scalacenter.github.io/bloop/), a
+project I've been recently working on, to achieve a **8x speedup in compile
+times**.
+
+The analysis and optimizations here presented can be migrated to any other
+Scala project that derives a lot of types at compile time or makes a heavy
+use of implicits and macros.
+
+After reading the blog post, you should understand:
+
+* How to use `scalac-profiling` to replicate a similar analysis.
+* Why Shapeless-based code is prone to slow compilation times if not used with
+  care.
+* How you can replicate a similar analysis of compilation times on code that
+  abuses implicit searches and macros.
+* How implicit search and macros interact in unexpected ways that hurt
+  productivity and how you can optimize their interaction.
+
+The most important take-away from this guide is that *you should not take
+slow Scala compile times for granted*. It's worth investigating why slow
+compiles happens as we can often reduce them on our side!
+
```

Comment:
I still think that this summary could use a lot of work. It's still presenting the takeaways in the wrong order. I would suggest using instead:

---

In this blog article, I'd like to introduce you to a new tool that we've
developed at the Scala Center called `scalac-profiling`, which aims to help you
better understand what is slowing down compilation in your project. 

It turns out that the use of Scala's implicits as well as macros can _greatly_
increase compilation times, depending on how they are used and how your codebase
is organized. Codebases that make use of libraries like Shapeless, or which rely
on typeclass derivation, may be particularly prone to these slow-downs. (As of
Scala 2.12, typeclass derivation is based on implicitly-triggered macro
expansions.) 

In this blog post, I will walk you through how to 
reduce these compile times with `scalac-profiling`. `scalac-profiling` is a
new Scala Center compiler plugin to complement my recent work on the compiler
statistics infrastructure merged in 2.12.5. In this guide, I use the plugin
to speed up a module of [Bloop](https://scalacenter.github.io/bloop/), a
project I've been recently working on, to achieve a **8x speedup in compile
times**.

The analysis and optimizations here presented can be migrated to any other
Scala project that make heavy use of typeclass deriving code, implicits, and/or macros.

After reading the blog post, you should understand:

* How to use `scalac-profiling` to analyze the impact of implicits and macros on compile times.
* Why typeclass deriving or Shapeless-based code is prone to slow compilation times if not used with care.
* How implicit search and macros interact in unexpected ways that can hurt performance and productivity and how you can optimize their interaction.

The most important take-away from this guide is that *you should not take
slow Scala compile times for granted*. It's worth investigating why slow
compiles happens as it is often possible to reduce slow compile times!